File: event/event.go
package event

import (
	"fmt"
	"strings"
	"time"

	"github.com/yagihash/fsw-calendar/fetcher"

	"google.golang.org/api/calendar/v3"
)

type Event struct {
	*calendar.Event
}

func New(date, start, end, title string) *Event {
	template := "%sT%s:00+09:00"

	return &Event{
		&calendar.Event{
			Summary: title,
			Start: &calendar.EventDateTime{
				DateTime: fmt.Sprintf(template, date, strings.TrimPrefix(start, "0")),
			},
			End: &calendar.EventDateTime{
				DateTime: fmt.Sprintf(template, date, strings.TrimPrefix(end, "0")),
			},
		},
	}
}

func NewFromDocEvent(d fetcher.DocEvent) *Event {
	return New(d.Date, d.Start, d.End, d.Title)
}

func (e *Event) Equals(c *Event) bool {
	es, err := time.Parse(time.RFC3339, e.Start.DateTime)
	if err != nil {
		panic(err)
	}

	ee, err := time.Parse(time.RFC3339, e.End.DateTime)
	if err != nil {
		panic(err)
	}

	cs, err := time.Parse(time.RFC3339, c.Start.DateTime)
	if err != nil {
		panic(err)
	}

	ce, err := time.Parse(time.RFC3339, c.End.DateTime)
	if err != nil {
		panic(err)
	}

	return e.Summary == c.Summary && es.Equal(cs) && ee.Equal(ce)
}

File: src/event/event.ts
export interface CalendarEventData {
  id?: string;
  summary: string;
  start: { dateTime: string };
  end: { dateTime: string };
}

export interface DocEvent {
  date: string;
  start: string;
  end: string;
  title: string;
}

function formatDateTime(date: string, time: string): string {
  return `${date}T${time}:00+09:00`;
}

export function newEvent(date: string, start: string, end: string, title: string): CalendarEventData {
  return {
    summary: title,
    start: { dateTime: formatDateTime(date, start) },
    end: { dateTime: formatDateTime(date, end) },
  };
}

export function newFromDocEvent(d: DocEvent): CalendarEventData {
  return newEvent(d.date, d.start, d.end, d.title);
}

export function eventsEqual(a: CalendarEventData, b: CalendarEventData): boolean {
  const as = new Date(a.start.dateTime).getTime();
  const ae = new Date(a.end.dateTime).getTime();
  const bs = new Date(b.start.dateTime).getTime();
  const be = new Date(b.end.dateTime).getTime();
  return a.summary === b.summary && as === bs && ae === be;
}
